### ISSUE TITLE
NUnit and TestCentric engines use different communication architectures

### ISSUE THREAD

**Issue by CharliePoole** [TestCentric/testcentric-gui#592]
### NUnit Compatibility Issue

__Category:__ Implementation
__Visibility:__ Engine Extenders and Internal
__Resolution:__ The approach in TestCentric will be ported to NUnit

#### Description

Both engines were limited by relying entirely on remoting. The two projets have taken different approaches to resolve this. NUnit is in the process of replacing remoting with a TCP protocol. TestCentric has moved the communication protocol, messaging and transport into a separate layer using a modified strategy pattern. Both remoting and TCP are currently in use and named pipes are also available but not included in the latest rlease.

It's not clear whether this is purely a matter of internal implementation or whether it may be visible to users at some point. That has to be looked at. There is a small possibility that a platform could come along, which requires use of a different pluggable protocol.

**Comment by CharliePoole** [comment]
@ChrisMaddock @jnm2 
I added this issue to the compatibility project because it seems to me that this is a pretty large ongoing divergence, which we should address. I had thought the particular direction of replacing Remoting had been abandoned and Chris had expressed an interest in looking at the alternative approach already released in the TestCetnric engine. But some recent PRs seem to be reviving the more monolithic approach of having only one comm protocol. This will be a rather large addition to the list of incompatibilities, so I'm hoping we can address it ahead of it's happening.

**Comment by jnm2** [comment]
We could keep the existing abstractions or introduce new ones. I thought it was too early to abstract. Do you have an implementation of a second agent protocol? That would take the guesswork out of the abstractions and possibly be something we could borrow in the engine.

**Comment by CharliePoole** [comment]
I had a named pipe implementation but it's no longer in the public tree. As you say, I needed to have at least two exemplars to be able to abstract things successfully. The remoting implementation is there but it's a degenerate implementation because the protocols, messages and transport are all hidden inside of .NET Remoting. 

If you look at the tree for the GUI, you'll find the stuff in the Communication subdirectory under both the engine and the engine.core assemblies. Most things are in core, but only the engine needs a proxy for TestAgency.

You are correct that it's next to impossible to get the abstraction right on the first cut. I'm no about my third published cut and it still may need work. But IME getting it done and getting it out is the best way to start iterating until it's right. BTW a large portion of the code reviews on the TestCentric project are done after merging. You'd be welcome to join in.

**Comment by jnm2** [comment]
Should we consider bringing TestCentric's existing work into the console rather than doing a copy of the work ourselves (https://github.com/nunit/nunit-console/pull/774)? (I'm a fan of getting out of doing work.) How would that work with copyright licensing?

Is a unification like this what is being proposed in this issue?

**Comment by CharliePoole** [comment]
@jnm2 Glad you like it!

TLDR; First decide if NUnit wants to change. Then deal with the mechanics.

------------------------------------------------------

I wanted this to be a place to discuss any and all incompatibilities without prejudging the solution, so I tried to keep the wording of each issue completely neutral. I'm only proposing that we figure out what we want to do.

That said, since you ask, I would prefer to see a communications layer where multiple options exist or can be added easily. I see adding named pipes (which I already did once) and eventually http: For named pipes, the same messages and protocols can be used. Not sure about http yet. First question would be to decide what we need for a .NET Core 1.1 agent to work. I currently run 1.1 tests under the 2.1 agent.

If it's decided to adopt this approach, I would like to work with you on implementing it.

Regarding licensing, my work is MIT licensed and I would contribute it in any case, so there's no problem. Any change in copyright would probably be best addressed with the 4.0 release, with  @ChrisMaddock taking the lead.

**Comment by CharliePoole** [comment]
@ChrisMaddock @jnm2 This issue had some discussion back in May but there's been no follow-up. As far as I can see we're each still going on our own paths. Can we come to any resolution here?

**Comment by ChrisMaddock** [comment]
This is one that needs discussing - in my mind it's one of the big ticket items for the NUnit Engine currently. I think both engines have the same goal here. Sorry I missed the previous discussions.

> Should we consider bringing TestCentric's existing work into the console rather than doing a copy of the work ourselves (nunit/nunit-console#774)? (I'm a fan of getting out of doing work.) 

@jnm2 - I absolutely agree with what you said 6 months ago. Would you be happy with that - given I know you've already done an amount of work on a new solution for the NUnit Engine? It does seem the most sensible solution, to save re-inventing the wheel.

Charlie - would you still be willing to port this over, given your comments last May? Apologies again I missed them before!

**Comment by CharliePoole** [comment]
@ChrisMaddock Absolutely! But let's hear from @jnm2 on this again first.

If we go with a minimalist 4.0, then this could be a 4.1 feature.

**Comment by jnm2** [comment]
Yes, I would be happy with that. It's very kind of Charlie to refer to the amount of work I've been doing as "going on a path" at all. 😊

**Comment by ChrisMaddock** [comment]
Great - thanks both! 😊

**Comment by ChrisMaddock** [comment]
@CharliePoole - are we happy for this to move to "awaiting action"?

**Comment by CharliePoole** [comment]
Waiting only for nunit/nunit-console#923 to be reviewed and merged to close this.
